terraform/hcl.go

```diff
@@ -0,0 +1,248 @@
+					Address:      addr,
+					Index:        i,
+					Mode:         "managed",
+					Type:         rv.Type,
+					Name:         rv.Name,
+					ProviderName: rv.Provider.Type,
+					Values:       cfg,
+				})
+			}
+			queries = append(queries, query.Resource{
+				Address:    addr,
+				Components: comps,
+			})
+		}
+	}
+
+	// Recursively extract resources from all child module calls.
+	for mk, mv := range mod.ModuleCalls {
+		p := joinPath(modPath, mv.SourceAddr)
+		child, diags := parser.LoadConfigDir(p)
+		if diags.HasErrors() {
+			return nil, fmt.Errorf(diags.Error())
+		}
+
+		body, ok := mv.Config.(*hclsyntax.Body)
+		if !ok {
+			return nil, fmt.Errorf("bad body")
+		}
+
+		// Extract variables from the module call block to pass down to the module.
+		vars := make(map[string]cty.Value)
+		for _, attr := range body.Attributes {
+			val, diags := attr.Expr.Value(evalCtx)
+			if diags != nil && diags.HasErrors() {
+				continue
+			}
+			vars[attr.Name] = val
+		}
+
+		nextEvalCtx := getEvalCtx(child, vars)
+
+		// If the module call contains a `providers` block, it should replace the implicit provider
+		// inheritance. Instead, a new map of parent to child providers is created.
+		// https://www.terraform.io/docs/language/modules/develop/providers.html#passing-providers-explicitly
+		var childProvs map[string]Provider
+		if len(mv.Providers) == 0 {
+			childProvs = providers
+		} else {
+			childProvs = make(map[string]Provider)
+			for _, p := range mv.Providers {
+				prov, ok := providers[p.InParent.String()]
+				if !ok {
+					continue
+				}
+				childProvs[p.InChild.String()] = prov
+			}
+		}
+
+		qs, err := extractHCLModule(childProvs, parser, p, mk, child, nextEvalCtx)
+		if err != nil {
+			return nil, err
+		}
+		queries = append(queries, qs...)
+	}
+
+	return queries, nil
+}
+
+// getEvalCtx returns the evaluation context of the given module with variable values set.
+func getEvalCtx(mod *configs.Module, vars map[string]cty.Value) *hcl.EvalContext {
+	// Set default values for undefined variables.
+	if vars == nil {
+		vars = make(map[string]cty.Value)
+	}
+	for vk, vv := range mod.Variables {
+		if _, ok := vars[vk]; !ok {
+			vars[vk] = vv.Default
+		}
+	}
+
+	evalCtx := &hcl.EvalContext{
+		Variables: map[string]cty.Value{
+			"var": cty.ObjectVal(vars),
+		},
+	}
+
+	// Set values of locals.
+	lm := make(map[string]cty.Value)
+	for lk, lv := range mod.Locals {
+		val, diags := lv.Expr.Value(evalCtx)
+		if diags != nil && diags.HasErrors() {
+			continue
+		}
+		lm[lk] = val
+	}
+	evalCtx.Variables["local"] = cty.ObjectVal(lm)
+
+	return evalCtx
+}
+
+// getBodyJSON gets all the variables in a JSON format of the actual representation
+func getBodyJSON(b *hclsyntax.Body, evalCtx *hcl.EvalContext) map[string]interface{} {
+	links := make(map[string]interface{})
+	for attrk, attrv := range b.Attributes {
+		val, _ := attrv.Expr.Value(evalCtx)
+		if !val.IsKnown() {
+			continue
+		}
+		switch val.Type() {
```

Comment:
No need to deal with maps or slices?

Reply:
IMO there's no need, I can't think of any resource that would use them for any cost-related attributes. For sure we don't support any ATM. Though there's `bool` missing that I will add.